### ISSUE TITLE
feature-kag

### ISSUE THREAD

**Issue by hi20260204-maker** [SKN-27th-FitLog/ScratchPaper#178]
## 🛩️ Description
그래프 DB 기본 틀 잡고 구현하기
- 아직 DB에는 미연결

**Comment by hi20260204-maker** [comment]
KAG / ETL 최근 변경 및 미해결 이슈 리포트
작성 기준: 최근 etl/meal, kag, FineDining 작업 내역

============================================================
1. 작업 범위 요약
============================================================

최근 작업은 크게 다음 범위로 진행되었다.

- etl/meal 산출 경로를 백엔드 코드 체계에 맞게 정리
- category_cd, shop_cd 저장 기준 정리
- process=cleansing 경로명을 process=cleaning으로 변경
- URL / 이미지 값을 DB 저장 직전에 HTML 태그 형태로 변환
- kag importer가 변경된 etl/meal Hive 경로를 읽도록 수정
- 레시피 재료 적재 정책을 "첫 번째 유효 레시피 기준"으로 변경
- 메뉴 품질 문제 분석
- 별도 프로젝트 FineDining을 생성하여 메뉴 정답 데이터 축적 기반 마련

주의:
- etl/it_news는 사용자가 명시적으로 "건드리면 안 된다"고 했기 때문에 변경하지 않았다.


============================================================
2. ETL 변경 사항
============================================================

2.1 etl/meal Hive 경로 변경
------------------------------------------------------------

기존에는 일부 경로에서 process=cleansing 형태를 사용했으나,
현재는 process=cleaning으로 변경했다.

기존:
  process=cleansing

변경:
  process=cleaning

또한 기존에는 category_cd=SC01처럼 업종 코드가 category_cd 파티션에
들어갈 수 있었으나, 백엔드 모델 기준에 맞춰 다음 구조로 변경했다.

기존 예시:
  process=cleaning/category_cd=SC01/...

변경 예시:
  process=cleaning/category_cd=CA01/shop_cd=SC01/...

의미:
  category_cd=CA01 : 맛집/음식점 카테고리
  shop_cd=SC01     : 한식
  shop_cd=SC02     : 일식
  shop_cd=SC03     : 중식
  shop_cd=SC04     : 양식
  shop_cd=SC05     : 카페
  shop_cd=SC06     : 퓨전

주요 변경 파일:
  etl/meal/src/core/storage/path_builder.py
  etl/meal/local_pipeline_runner.py


2.2 category_cd, shop_cd 저장 기준 변경
------------------------------------------------------------

백엔드 모델 기준에 맞춰 DB 저장 시 다음 기준을 적용했다.

  maps.category_cd     = CA01
  crawling.category_cd = CA01
  shop.shop_cd         = SCxx

즉, 맛집이라는 상위 카테고리는 CA01로 저장하고,
한식/일식/중식/카페 같은 업종 구분은 shop_cd로 저장한다.

주요 변경 파일:
  etl/meal/src/core/constants.py
  etl/meal/src/core/repository/store_repository.py
  etl/meal/src/projects/process/stage3_validation_normalization.py
  etl/meal/src/projects/save/stage4_load.py


2.3 URL / 이미지 HTML 태그 변환
------------------------------------------------------------

사용자가 요청한 대로, ETL에서 수집한 URL과 이미지 URL은
DB 적재 직전에 HTML 태그 형태로 변환하도록 처리했다.

URL 저장 형태:
  <a href="https://news.example.com/post/1">PyTorch 토론 요약</a>

이미지 저장 형태:
  <img src="https://cdn.example.com/screenshot.png" alt="토론 스크린샷"/>

현재 etl/meal에서는 다음과 같이 적용된다.

- crawling.article_url
  원본 URL을 <a href="...">...</a> 형태로 저장

- images.image_url
  원본 이미지 URL을 <img src="..." alt="..."/> 형태로 저장

중요한 점:
- 이 처리는 ai 폴더가 아니라 etl/meal의 DB 적재 직전 단계에서 처리된다.
- ai/postmake_pipeline 쪽 잘못된 변경은 제거했고,
  현재 ai 쪽 실제 내용 diff는 없는 상태로 확인했다.

주요 변경 파일:
  etl/meal/src/projects/save/stage4_load.py


2.4 레시피 수집 흐름
------------------------------------------------------------

etl/meal/local_pipeline_runner.py 기준 실행 흐름은 다음과 같다.

  Crawl -> Process -> Recipe -> Save

즉, 로컬 파이프라인을 실행하면 다음 단계가 순차적으로 진행된다.

1. DiningCode 크롤링
2. 후보 데이터 파싱
3. 정규화
4. 레시피 수집
5. PostgreSQL 저장

단, KAG Neo4j 적재는 자동으로 이어지지 않는다.
KAG 적재는 별도 명령으로 실행해야 한다.


============================================================
3. KAG 변경 사항
============================================================

3.1 KAG importer 경로 변경
------------------------------------------------------------

kag importer가 변경된 etl/meal 산출 경로를 읽도록 수정했다.

Meal cleaning 경로:
  process=cleaning/category_cd=CA01/shop_cd=*/year=*/month=*/day=*/status=success/validation_normalization_*.jsonl

Meal recipe 경로:
  process=recipe/category_cd=CA01/shop_cd=*/year=*/month=*/day=*/status=success/recipe_collection_*.jsonl

IT news 경로는 변경하지 않았다.

IT news 경로:
  process=cleaning/category_cd=IC02/year=*/month=*/day=*/status=success/*.csv

주요 변경 파일:
  kag/src/kag_graph/hive_importer.py
  kag/tests/test_hive_importer.py


3.2 레시피 재료 적재 정책 변경
------------------------------------------------------------

기존 KAG importer는 같은 recipe_search_keyword에 여러 레시피가 있으면
재료를 병합했다.

기존 방식:
  제육볶음 recipe 1 재료 + 제육볶음 recipe 2 재료 + ...

변경 후 방식:
  같은 recipe_search_keyword에서 첫 번째로 재료가 있는 레시피 하나만 사용

예시 입력:

[
  {
    "recipe_search_keyword": "제육볶음",
    "recipe_url": "https://www.10000recipe.com/recipe/empty",
    "ingredients": []
  },
  {
    "recipe_search_keyword": "제육볶음",
    "recipe_url": "https://www.10000recipe.com/recipe/first",
    "ingredients": [
      {"ingredient_name": "돼지고기"},
      {"ingredient_name": "양파"}
    ]
  },
  {
    "recipe_search_keyword": "제육볶음",
    "recipe_url": "https://www.10000recipe.com/recipe/second",
    "ingredients": [
      {"ingredient_name": "대파"}
    ]
  }
]

KAG에 들어가는 결과:

{
  "제육볶음": [
    {
      "ingredient_name": "돼지고기",
      "recipe_url": "https://www.10000recipe.com/recipe/first"
    },
    {
      "ingredient_name": "양파",
      "recipe_url": "https://www.10000recipe.com/recipe/first"
    }
  ]
}

즉, 빈 레시피는 건너뛰고 첫 번째 유효 레시피의 재료만 사용한다.

주요 변경 파일:
  kag/src/kag_graph/hive_importer.py
  kag/tests/test_recipe_ingredient_policy.py
  kag/tests/test_hive_importer.py


3.3 KAG 재료 source/confidence 상수화
------------------------------------------------------------

기존에는 다음 값들이 코드 내부에 직접 들어가 있었다.

  RECIPE_10000
  CSV_FALLBACK
  0.9
  0.5

이를 상수로 분리했다.

  INGREDIENT_SOURCE_RECIPE = "RECIPE_10000"
  INGREDIENT_SOURCE_CSV_FALLBACK = "CSV_FALLBACK"
  INGREDIENT_CONFIDENCE_RECIPE = 0.9
  INGREDIENT_CONFIDENCE_CSV_FALLBACK = 0.5

의미:
- 레시피 기반 재료는 RECIPE_10000, confidence 0.9
- CSV fallback 재료는 CSV_FALLBACK, confidence 0.5


============================================================
4. 테스트 및 검증 결과
============================================================

4.1 ETL 테스트
------------------------------------------------------------

실행한 주요 테스트:
  etl/meal/tests/test_stage4_backend_category_contract.py

결과:
  4 passed

검증 내용:
- maps.category_cd = CA01
- shop.shop_cd = SCxx
- crawling.article_url이 <a href="...">...</a> 형태로 저장됨
- images.image_url이 <img src="..." alt="..."/> 형태로 저장됨


4.2 KAG 테스트
------------------------------------------------------------

실행한 주요 테스트:
  kag/tests/test_recipe_ingredient_policy.py
  kag/tests/test_hive_importer.py::test_build_recipe_ingredient_map_uses_first_recipe_ingredients_by_keyword
  kag/tests/test_hive_importer.py::test_build_meal_restaurant_statements_uses_recipe_ingredients_over_csv
  kag/tests/test_hive_importer.py::test_build_meal_restaurant_statements_falls_back_to_csv_when_no_recipe
  kag/tests/test_hive_importer.py::test_collect_current_hive_statements_reads_recipe_files_from_meal_root

결과:
  5 passed

검증 내용:
- KAG가 meal recipe 새 경로를 읽음
- 첫 번째 유효 레시피만 재료로 사용
- 레시피 재료가 있으면 CSV fallback을 사용하지 않음
- 레시피 재료가 없으면 CSV fallback 사용
- meal cleaning 산출물과 recipe 산출물을 함께 읽어 statement 생성


4.3 컴파일 검증
------------------------------------------------------------

실행한 검증:
  py_compile kag/src/kag_graph/hive_importer.py

결과:
  passed


============================================================
5. 아직 해결하지 못한 부분
============================================================

5.1 KAG Neo4j 실제 적재는 자동이 아님
------------------------------------------------------------

현재 상태에서 etl/meal 파이프라인은 PostgreSQL 저장까지 수행한다.

현재 구조:
  ETL meal crawling/process/recipe/save -> PostgreSQL 저장

KAG 적재는 별도 명령 필요:
  .\.venv\Scripts\python.exe kag\scripts\import_hive_neo4j.py --mode current-hive --meal-root etl\meal\_temp_lake --it-news-root etl\it_news

미해결 사항:
- ETL 완료 후 KAG importer 자동 실행 트리거 없음
- AWS/EventBridge/Lambda/Scheduler 같은 자동화는 아직 구현하지 않음
- 로컬에서는 수동 명령으로 순차 실행해야 함


5.2 etl/it_news는 변경하지 않음
------------------------------------------------------------

사용자가 명시적으로 etl/it_news 폴더는 건드리면 안 된다고 했기 때문에
변경하지 않았다.

현재 KAG importer는 기존 it_news 경로를 유지한다.

  process=cleaning/category_cd=IC02/year=*/month=*/day=*/status=success/*.csv

미해결 사항:
- it_news 산출물이 실제 최신 실행 결과와 완전히 맞는지는 별도 실행 검증 필요
- 단, 코드 변경은 하지 않음


5.3 메뉴 품질 문제
------------------------------------------------------------

현재 메뉴 수집 결과에서 다음 문제가 확인되었다.

예시:
- 카페
- 디저트

분석 결과:

- 카페
  가격 4000
  설명 있음
  실제 음료 메뉴로 볼 수 있음
  무조건 제거하면 안 됨

- 디저트
  가격 0
  설명 없음
  메뉴 카테고리/섹션명일 가능성이 높음
  DROP 후보

현재 문제:

  Stage2 DiningCodeParser.parse_menus()
  -> 메뉴 목록을 그대로 수집

  Stage3 MenuNormalizer
  -> LOW_CONFIDENCE여도 메뉴 배열에 유지

  Stage4 Load
  -> menus 배열 전체를 DB menu 테이블에 저장

즉, 정규화 실패 여부와 무관하게 메뉴가 DB까지 흘러갈 수 있다.

미해결 사항:
- 메뉴 품질 필터는 아직 SK-Connect에 실제 적용하지 않음
- KEEP / REVIEW / DROP 정답 데이터 기반으로 추후 반영 예정
- 무리한 룰 기반 삭제는 실제 메뉴를 삭제할 위험이 있어 보류


5.4 주재료 판별은 제외
------------------------------------------------------------

이전 논의에서 "주재료 판별"은 현재 범위에서 제외하기로 했다.

이유:
- 된장, 고추장, 장 같은 재료는 음식에 따라 주재료일 수도 있고 양념일 수도 있음
- 단순 키워드 룰로 판별하면 하드코딩이 심해짐
- 정확히 하려면 정답 데이터, 모델, 또는 fine-tuning이 필요함

현재 정책:
  주재료 판별하지 않음
  첫 번째 유효 레시피의 재료 전체를 사용

미해결 사항:
- 주요 재료 / 부재료 / 양념 / 고명 분리는 아직 하지 않음
- 추후 FineDining 정답 데이터가 쌓이면 별도 모델 또는 라벨 기준으로 처리 예정


5.5 전체 KAG 테스트 일부 실패
------------------------------------------------------------

KAG 전체 테스트를 시도했을 때 일부 실패가 있었다.

원인:
1. repo root에서 실행할 경우 fixture 상대 경로 문제 발생
2. pytest 임시 디렉터리 권한 문제 발생

확인된 점:
- 이번 변경과 직접 관련된 KAG 테스트는 통과
- 전체 테스트 실패는 이번 변경 로직 자체의 실패라기보다 실행 환경/경로/권한 문제에 가까움

미해결 사항:
- KAG 전체 테스트를 안정적으로 실행할 수 있도록 pytest fixture 경로와 temp dir 설정 정리가 필요함


============================================================
6. FineDining 별도 프로젝트 생성
============================================================

SK-Connect 밖에 별도 프로젝트를 생성했다.

경로:
  C:\dev\Project\FineDining

생성 구조:

  C:\dev\Project\FineDining
  ├─ README.md
  ├─ data
  │  ├─ raw_menu_samples.jsonl
  │  ├─ menu_classification_labels.jsonl
  │  ├─ ingredient_labels.jsonl
  │  └─ schema.md
  └─ scripts
     └─ export_menu_samples.py

역할:
- SK-Connect의 메뉴 품질을 개선하기 위한 정답 데이터 축적 프로젝트
- 사람이 KEEP / REVIEW / DROP 라벨링
- 추후 SK-Connect 메뉴 필터, LLM classifier, fine-tuning 데이터셋에 활용

현재 추출 결과:
  raw_menu_samples.jsonl
  29379개 메뉴 샘플 추출 완료

주의:
- FineDining 데이터는 아직 KAG에 반영하지 않음
- FineDining 라벨 데이터도 아직 비어 있음
- 현재는 샘플 추출 및 스키마 준비 단계


============================================================
7. 현재 상태 요약
============================================================

완료됨:
- etl/meal Hive 경로를 category_cd=CA01/shop_cd=SCxx 구조로 변경
- process=cleansing을 process=cleaning으로 변경
- DB 저장 시 category_cd=CA01, shop_cd=SCxx 기준 반영
- URL / 이미지 HTML 태그 변환을 DB 적재 직전에 적용
- KAG importer가 meal 새 경로를 읽도록 수정
- KAG 레시피 재료 정책을 첫 번째 유효 레시피 기준으로 변경
- 관련 ETL/KAG 테스트 추가 및 통과
- FineDining 별도 프로젝트 생성
- SK-Connect 메뉴 샘플 29379개를 FineDining으로 추출

아직 안 됨:
- KAG Neo4j 자동 적재 트리거
- etl/it_news 변경
- 메뉴 품질 필터를 SK-Connect에 실제 적용
- FineDining 라벨 데이터 작성
- FineDining 라벨을 SK-Connect/KAG에 반영
- 주재료/부재료/양념 판별
- 전체 KAG 테스트 환경 정리
